### PR TITLE
Refactor wRPC

### PR DESCRIPTION
**Pull request by Suika-Kong**
1. Added comments(usage of APIs).
2. Split modules.
3. Redesigned the API, now `call` returns a `future` object to wait for the response. And added `call_noreturn` to drop the `future`. This also makes sure the error of the call is not ignored, and a response to a call that is not initiated by this peer will be an error.
4. Styling, including localized variables.
5. Remove unused code(channel). Another feature at later PR will do the objectives of this feature.

Fix FT-2942

### CODE REVIEW

File: kong/tools/wrpc/future.lua
Comment (chronolaw) [inline kong/tools/wrpc/future.lua:19]: We use snake_case, not camelCase, this should be `drop_wait`.

**Comment by Suika-Kong** [comment]
Once this PR is merged, I will try to change the interface more aggressively.
1. Manage reconnection inside the wRPC, so config service need not take care of it;
2. Add APIs to PDK to access the wRPC connection

**Review by javierguerragiraldez — CHANGES_REQUESTED** [review]
replacing pending IDs with futures is a good improvement.  function naming is still off.

File: kong/tools/wrpc/future.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/future.lua]: Fixed.

File: kong/clustering/wrpc_data_plane.lua
Comment (chronolaw) [inline kong/clustering/wrpc_data_plane.lua:222]: `void_call` is a bit strange, could we change it to a better name?

File: kong/tools/wrpc/future.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/future.lua:14]: I'm trying to tell this is an internal API that should not be called by users of wRPC. Maybe I could just explicitly say it.

File: kong/clustering/wrpc_data_plane.lua
Comment (Suika-Kong) [reply kong/clustering/wrpc_data_plane.lua:222]: how about `call_no_return`?

File: kong/tools/wrpc/threads.lua
Comment (kikito) [inline kong/tools/wrpc/threads.lua:63]: This looks dangerously close to an active waiting. What would be the downside of sleeping for longer instead of 0 (perhaps some configurable value)?

File: kong/tools/wrpc/future.lua
Comment (chronolaw) [inline kong/tools/wrpc/future.lua:77]: What is the meaning of `_t` in `response_t`? in C , `_t` means a type or struct.

File: kong/tools/wrpc/message.lua
Comment (chronolaw) [inline kong/tools/wrpc/message.lua:108]: Is `--`  a good desc for error? Is it defined in spec?

File: kong/tools/wrpc/message.lua
Comment (chronolaw) [inline kong/tools/wrpc/message.lua:44]: localize `tonumber`

File: kong/tools/wrpc/future.lua
Comment (kikito) [inline kong/tools/wrpc/future.lua:71]: perhaps I am missing some context here. I am having a hard time understanding what a "future" is supposed to represent here, and I am also confused about what the methods do.

Is this similar to the concept of Promise in Node.js, or somewhere else? If this is based on something, I suggest copying the method names from there.

File: kong/tools/wrpc/threads.lua
Comment (javierguerragiraldez) [reply kong/tools/wrpc/threads.lua:42]: i think the logic is correct and more readable, but depending on some decoding flags, those fields (`ack` and `deadline`) could arrive here with `0` as the "undefined" value, and thus not falsy.  that's because of the "not exactly static-type" semantics of protobuf.

File: kong/tools/wrpc/threads.lua
Comment (javierguerragiraldez) [reply kong/tools/wrpc/threads.lua:63]: `step()` does block on I/O, so the sleep might not be that important.  perhaps only do it if there was an error?  maybe even move it (the sleep) inside `step()`?

File: kong/tools/wrpc/future.lua
Comment (javierguerragiraldez) [reply kong/tools/wrpc/future.lua:71]: i've seen promises and futures used mostly interchangeably.  i'm not sure which i'd prefer.  I thought SICP used "futures", but reviewing it it mostly calls them "expressions with lazy evaluation", and only informally mentions that they represent a "promise".

according to [wikipedia](https://en.wikipedia.org/wiki/Futures_and_promises):

> The terms future, promise, delay, and deferred are often used interchangeably, although some differences in usage between future and promise are treated below. Specifically, when usage is distinguished, a future is a read-only placeholder view of a variable, while a promise is a writable, [single assignment](https://en.wikipedia.org/wiki/Single_assignment) container which sets the value of the future.

in C++11 there are both `promise` and `future` objects, [where](https://www.cplusplus.com/reference/future/promise/):

> - The promise object is the asynchronous provider and is expected to set a value for the shared state at some point.
> - The future object is an asynchronous return object that can retrieve the value of the shared state, waiting for it to be ready, if necessary.

so, the producer constructs a `promise` and from it a `future`.  the "user" gets the `future` and the background process writes to the `promise`, making the `future` ready (an waking anybody that might've been waiting for it)

so.... i'd be (very) slightly inclined to call this a future.

File: kong/clustering/wrpc_data_plane.lua
Comment (chronolaw) [reply kong/clustering/wrpc_data_plane.lua]: How about `async_call`?

File: kong/tools/wrpc/threads.lua
Comment (chronolaw) [inline kong/tools/wrpc/threads.lua:70]: I think we should avoid `else`, use `if not then goto continue` to reduce the nested layer.

File: kong/tools/wrpc/utils.lua
Comment (chronolaw) [inline kong/tools/wrpc/utils.lua:3]: is `-- luacheck: ignore` necessary? Could we remove this comments?

Comment (javierguerragiraldez) [reply kong/tools/wrpc/utils.lua:3]: here `table.insert()` is only used in the "append" form, which is compiled to exactly the same IR as `t[#t+1] = x`  so it could just be avoided.

File: kong/tools/wrpc/future.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/future.lua:30]: Comment updated.

File: kong/tools/wrpc/future.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/future.lua:77]: table.

File: kong/tools/wrpc/message.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/message.lua:108]: Not sure. This is simply copied from the origin code.

File: kong/tools/wrpc/threads.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/threads.lua:42]: I will leave it unchanged but add comments to explain this.

File: kong/tools/wrpc/threads.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/threads.lua:63]: I will do this:

```suggestion
    while not exiting() and not wrpc_peer.closing do
      local ok = step(wrpc_peer)
      -- something wrong with this step
      -- let yield instead of retry immediately
      if not ok then
        sleep(0)
      end
    end
```

File: kong/tools/wrpc/threads.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/threads.lua:84]: Done

File: kong/tools/wrpc/threads.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/threads.lua:70]: I will just make it 2 ifs.

File: kong/tools/wrpc/message.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/message.lua:33]: It's due to the design of wRPC, and the function is doing only 1 thing. To split it makes it harder to read.

The function is about 70 lines long and can be read without scrolling down. It has a lot of empty lines and comments, and line bending. I don't think this function is too large.

**Comment by Suika-Kong** [comment]
I also fixed a bug, that when an RPC returns an empty message, the payload itself will be "nil" and cause a crash.
The logic that causes the problem is nearly the same as the origin, but somehow it did not make tests fail every time.

File: kong/tools/wrpc/message.lua
Comment (fffonion) [inline kong/tools/wrpc/message.lua:65]: the `error` parameter seems not set in the `send_error` function. we should also check return of this function. consider doing something like this:
```lua
return send_error(wrpc_peer, payload, {
      etype = "ERROR_TYPE_INVALID_SERVICE",
      description = "Invalid service (or rpc)",
    })
```
then in send_error, do
```lua
local function send_error(wrpc_peer, payload, error)
  local ok, err = wrpc_peer:send_payload({
    mtype = "MESSAGE_TYPE_ERROR",
    error = error,
    svc_id = payload.svc_id,
    rpc_id = payload.rpc_id,
    ack = payload.seq,
  })
  if not ok then
    return ok, err
  end
  return nil, err.description
end
```

File: kong/tools/wrpc/message.lua
Comment (fffonion) [inline kong/tools/wrpc/message.lua:115]: check return of this function

File: kong/tools/wrpc/message.lua
Comment (fffonion) [inline kong/tools/wrpc/message.lua:143]: check payload.ack

File: kong/tools/wrpc/message.lua
Comment (fffonion) [inline kong/tools/wrpc/message.lua:107]: check return and fix parameter passed in

File: kong/tools/wrpc/message.lua
Comment (fffonion) [inline kong/tools/wrpc/message.lua:126]: check return and fix parameter passed in

**Comment by Suika-Kong** [comment]
Those modules are split and copied from the origin code. I'll refactor them later. To review those code, you can ignore them as they are not changed:
1. message.lua (except for extracting `send_error`;
2. queue.lua(no change except for rename Queue to `_M` and `_MT`);
3. threads.lua(except for add comments)

File: kong/tools/wrpc/message.lua
Comment (fffonion) [inline kong/tools/wrpc/message.lua:78]: wrap this into a function of response_future would be a good idea.

File: kong/tools/wrpc/message.lua
Comment (fffonion) [inline kong/tools/wrpc/message.lua:169]: need an error handling for else of "if rpc and rpc.error_handler", just a log for now is okay

File: kong/tools/wrpc/message.lua
Comment (fffonion) [reply kong/tools/wrpc/message.lua:72]: yes, i agree `local ack = payload.ack or 0` is sufficient here, let's remove the `tonumber` then.

File: kong/tools/wrpc/message.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/message.lua:81]: And yes, we can check expire time in the future itself as it is aware of the expire time( and it handles expiration when waiting).

File: kong/tools/wrpc/message.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/message.lua:65]: I'm missing the payload argument.
Suggestion applied.

File: kong/tools/wrpc/message.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/message.lua:143]: It misses the case when ack<0. But ack must be either number or nil.

File: kong/tools/wrpc/message.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/message.lua:115]: Ditto.

File: kong/tools/wrpc/message.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/message.lua:78]: done

File: kong/tools/wrpc/message.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/message.lua:169]: done

File: kong/tools/wrpc/message.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/message.lua:149]: I will leave this to later refactoring.

File: kong/tools/wrpc/message.lua
Comment (fffonion) [inline kong/tools/wrpc/message.lua:84]:
```suggestion
  return wrpc_peer:send_payload({
```

File: kong/tools/wrpc/message.lua
Comment (fffonion) [inline kong/tools/wrpc/message.lua:93]:
```suggestion
```

**Review by fffonion — COMMENTED** [review]
minor issue

File: kong/tools/wrpc/proto.lua
Comment (javierguerragiraldez) [inline kong/tools/wrpc/proto.lua:159]: this now supports a single argument, better rename to `:encode_arg`

File: kong/tools/wrpc/init.lua
Comment (javierguerragiraldez) [inline kong/tools/wrpc/init.lua:184]: make this single-argument too (if `:encode_args()` no longer supports multiple args)

File: kong/tools/wrpc/init.lua
Comment (javierguerragiraldez) [inline kong/tools/wrpc/init.lua:314]: this line is like this to support multiple args.  if that feature is removed, this line should be much simpler

**Review by javierguerragiraldez — CHANGES_REQUESTED** [review]
multiple arguments was discussed and decided to be removed to simplify the code, but i still see several places where the code clearly tries to support it (and fails, because other parts now assume single arg).

basically search for every `table_unpack` and `...`  all of them should go.

File: kong/tools/wrpc/proto.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/proto.lua:159]: I think it's still `args`, but all arguments for 1 call.

File: kong/tools/wrpc/init.lua
Comment (Suika-Kong) [reply kong/tools/wrpc/init.lua:184]: I was planning to do this at another PR. But I think it's OK to include this.

**Review by javierguerragiraldez — APPROVED** [review]

**Review by fffonion — APPROVED** [review]

**Pull request merged**
